[JSC] Iterator Helpers methods should not iterate an array

sosukesuzuki opened the pull request:
#### ea7b8287cc3f7f5b98cf261bcca1ea12829a18bf
<pre>
[JSC] Iterator Helpers methods should not iterate an array
<a href="https://bugs.webkit.org/show_bug.cgi?id=279876">https://bugs.webkit.org/show_bug.cgi?id=279876</a>

Reviewed by Alexey Shvayka.

We implemented a function called `forEachInIteratorProtocol` to implement
`Iterator.prototype.toArray`[1]. As the name suggests, this function iterates over objects that
conform to the iterator protocol, but it also iterates over fast arrays.

The methods of Iterator Helpers use the `GetIteratorDirect` abstract operation[2] to directly get
the object&apos;s `next` property. Unlike the `GetIterator` abstract operation[3], it does not get the
`@@iterator`.

This patch changes the `forEachInIteratorProtocol` function to stop iterating over arrays. It also
removes the fast path for fast arrays in `Iterator.prototype.toArray`[4].

[1]: <a href="https://commits.webkit.org/283381@main">https://commits.webkit.org/283381@main</a>
[2]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-getiteratordirect">https://tc39.es/proposal-iterator-helpers/#sec-getiteratordirect</a>
[3]: <a href="https://tc39.es/ecma262/#sec-getiterator">https://tc39.es/ecma262/#sec-getiterator</a>
[4]: <a href="https://commits.webkit.org/283381@main">https://commits.webkit.org/283381@main</a>

* JSTests/microbenchmarks/iterator-prototype-toArray-for-array.js: Removed.
* JSTests/stress/iterator-prototype-forEach.js:
(sameArray):
(const.validIter): Deleted.
(const.invalidCallback.of.invalidCallbacks.shouldThrow): Deleted.
* JSTests/stress/iterator-prototype-toArray-empty.js: Removed.
* JSTests/stress/iterator-prototype-toArray.js:
(sameArray):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIteratorProtocol):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/283933@main">https://commits.webkit.org/283933@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5936ebd3b8c6a1754c3073343b1036aa33747fe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67715 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20354 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71769 "Built successfully") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54896 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18662 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54192 "Passed tests") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43187 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58548 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39861 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15954 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17214 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60832 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61827 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16296 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73467 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66962 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11678 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15604 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61636 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11713 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58617 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61683 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9521 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3137 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88730 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10315 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42904 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15649 "Found 8 new JSC stress test failures: wasm.yaml/wasm/function-tests/many-arguments-to-function.js.wasm-no-cjit, wasm.yaml/wasm/fuzz/memory.js.wasm-bbq, wasm.yaml/wasm/fuzz/memory.js.wasm-slow-memory, wasm.yaml/wasm/stress/js-to-wasm-many-i64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/js-wasm-call-many-return-types-on-stack-no-args.js.wasm-eager-jettison, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-collect-continuously, wasm.yaml/wasm/stress/tail-call.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43980 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45167 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43719 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shvaikalesh reviewed (commented):
Looks good but could you please expand the ChangeLog? This part:

> According to the specification, this behavior is unexpected

Shall we always `[[Get]]` the iterator & its `next` method?

sosukesuzuki committed:
[JSC] Iterator Helpers methods should not iterate an array
https://bugs.webkit.org/show_bug.cgi?id=279876

Reviewed by Alexey Shvayka.

We implemented a function called `forEachInIteratorProtocol` to implement
`Iterator.prototype.toArray`[1]. As the name suggests, this function iterates over objects that
conform to the iterator protocol, but it also iterates over fast arrays.

The methods of Iterator Helpers use the `GetIteratorDirect` abstract operation[2] to directly get
the object's `next` property. Unlike the `GetIterator` abstract operation[3], it does not get the
`@@iterator`.

This patch changes the `forEachInIteratorProtocol` function to stop iterating over arrays. It also
removes the fast path for fast arrays in `Iterator.prototype.toArray`[4].

[1]: https://commits.webkit.org/283381@main
[2]: https://tc39.es/proposal-iterator-helpers/#sec-getiteratordirect
[3]: https://tc39.es/ecma262/#sec-getiterator
[4]: https://commits.webkit.org/283381@main

* JSTests/microbenchmarks/iterator-prototype-toArray-for-array.js: Removed.
* JSTests/stress/iterator-prototype-forEach.js:
(sameArray):
(const.validIter): Deleted.
(const.invalidCallback.of.invalidCallbacks.shouldThrow): Deleted.
* JSTests/stress/iterator-prototype-toArray-empty.js: Removed.
* JSTests/stress/iterator-prototype-toArray.js:
(sameArray):
* Source/JavaScriptCore/runtime/IteratorOperations.h:
(JSC::forEachInIteratorProtocol):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/283933@main
